test/unit/spec/util/backoff.js

```diff
@@ -0,0 +1,123 @@
+'use strict';
+
+const DefaultBackoff = require('../../../../lib/util/backoff');
+const sinon = require('sinon');
+
+describe('Backoff', () => {
+  describe('Function Call', () => {
+    let options = {};
+    let fn;
+    let backoff;
+    let fakeTimer;
+    beforeEach(() => {
+      fn = sinon.spy();
+      backoff = new DefaultBackoff(options);
+      fakeTimer = sinon.useFakeTimers();
+    });
+
+    afterEach(() => {
+      fn.resetHistory();
+      backoff.reset();
+      fakeTimer.restore();
+    });
+
+    it('should call the function on start', () => {
+      backoff.backoff(fn);
+      fakeTimer.tick(100);
+      backoff.reset();
+      sinon.assert.calledOnce(fn);
+    });
+
+    it('should increase the duration exponentially', () => {
+      backoff.backoff(fn);
+      fakeTimer.tick(110);
+      backoff.backoff(fn);
+      fakeTimer.tick(210);
+      backoff.backoff(fn);
+      fakeTimer.tick(410);
+      sinon.assert.calledThrice(fn);
+    });
+
+    it('should reset the duration', () => {
+      backoff.backoff(fn);
+      fakeTimer.tick(110);
+      backoff.reset();
+      fakeTimer.tick(210);
+      fakeTimer.tick(410);
+      sinon.assert.calledOnce(fn);
```

Comment:
why is it getting called just once here? shouldn't it get called thrice?

Reply:
No, because we only called `backoff` once. 

Previously though, it also would have been one because it should not have called `fn` after we call reset unless we start `backoff` again.